src/components/pages/dashboard-pages/Students.jsx

```diff
@@ -1,11 +1,327 @@
+  const { userData } = useAuth();
+  const [students, setStudents] = useState([]);
+  const [selectedId, setSelectedId] = useState();
+  const [selectedStudent, setSelectedStudent] = useState();
+  const [studentsError, setStudentsError] = useState({});
+  const [isLoading, setIsLoading] = useState(false);
+  const [studentLessons, setStudentLessons] = useState([]);
+  const [classTitles, setClassTitles] = useState([]);
+
+  useEffect(() => {
+    setIsLoading(true);
+    const getTeacherStudents = async () => {
+      const token = userData.token;
+      try {
+        const allUsers = await getAllUsersInfo(token);
+        const myStudents = userData.myStudents;
+        const filteredStudents = allUsers.filter((user) =>
+          myStudents.includes(user._id)
+        );
+        if (filteredStudents.length === 0) {
+          setIsLoading(false);
+          setStudentsError({
+            noStudentsError: `You haven't added any students yet.`,
+          });
+        } else {
+          setStudents(filteredStudents);
+          setSelectedId(filteredStudents[0]._id);
+          setStudentsError({ fetchError: '', noStudentsError: '' });
+          setIsLoading(false);
+        }
+      } catch (error) {
+        console.error('Error fetching students data:', error);
+        setStudentsError({
+          fetchError: 'Failed to fetch students. Please try again later.',
+        });
+      }
+    };
+    getTeacherStudents();
+  }, [userData]);
+
+  useEffect(() => {
+    const getClassesInfo = async () => {
+      try {
+        const response = await getClassesData();
+        const allClasses = response.classes;
+        const myClassIds = userData.myClasses.map((id) => id);
+        const filteredClasses = allClasses.filter((classes) =>
+          myClassIds.includes(classes._id)
+        );
+        const myClassesTitleAndId = filteredClasses.map((classItem) => ({
+          title: classItem.classTitle,
+          id: classItem._id,
+        }));
+        setClassTitles(myClassesTitleAndId);
+      } catch (error) {
+        console.error('Error fetching lessons data:', error);
+      }
+    };
+    getClassesInfo();
+  }, [userData]);
+
+  useEffect(() => {
+    const getLessonsInfo = async () => {
+      if (!selectedId) return;
+
+      const token = userData.token;
+      try {
+        const response = await getAllStudentLessons(token, selectedId);
+        const allStudentLessons = response.data.lessons;
+        // Group lessons by classId
+        const groupedLessons = allStudentLessons.reduce((acc, lesson) => {
+          const { classId } = lesson;
+          if (!acc[classId]) {
+            acc[classId] = [];
+          }
+          acc[classId].push(lesson);
+          return acc;
+        }, {});
+        setStudentLessons(groupedLessons);
+      } catch (error) {
+        console.error('Error fetching lessons data:', error);
+      }
+    };
+
+    getLessonsInfo();
+  }, [selectedId]);
+
+  useEffect(() => {
```

Comment:
this useEffect is unnecessary. because your state `selectedStudent` based on 2 other states: selectedId, students, you can avoid using useEffect and useState for `selectedStudent`.

use derivative variable instead: const selectedStudent  = selectedId  ? students.find(
        (student) => student._id === selectedId
      ) : null

Reply:
Thank you!
Done